src/main/java/org/opensearch/knn/index/vectorvalues/VectorValueExtractorStrategy.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.opensearch.knn.index.vectorvalues;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.ByteVectorValues;
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.BytesRef;
+import org.opensearch.knn.index.VectorDataType;
+import org.opensearch.knn.index.codec.util.KNNVectorSerializer;
+import org.opensearch.knn.index.codec.util.KNNVectorSerializerFactory;
+
+import java.io.IOException;
+
+/**
+ * Provides different strategies to extract the vectors from different {@link KNNVectorValuesIterator}
+ */
+interface VectorValueExtractorStrategy {
+    /**
+     * Extract Vector based on the vector datatype and vector values iterator.
+     * @param vectorDataType {@link VectorDataType}
+     * @param vectorValuesIterator {@link KNNVectorValuesIterator}
+     * @return vector
+     * @param <T> could be of type float[], byte[]
+     * @throws IOException exception during extracting the vectors
+     */
+    <T> T extract(VectorDataType vectorDataType, KNNVectorValuesIterator vectorValuesIterator) throws IOException;
```

Comment:
I still against this pattern here. 
It is like you have to go inside extract method and check what it returns based on your parameter. 
It would be much easier to read code by having
```
byte[] extractBinaryVector(KNNVectorValuesIterator vectorValuesIterator)
byte[] extractByteVector(KNNVectorValuesIterator vectorValuesIterator)
float[] extractFloatVector(KNNVectorValuesIterator vectorValuesIterator)
```

Reply:
Lets keep it like this for now, but I think this way we avoid a lot of extraction methods/classes per IteratorType and then per vector datatype. As use of generics is quite new in k-NN plugin but a prominent thing in Java. Its more around getting used to in reading the code with generics

Reply:
It is not about getting used to generic. Just this is not a best practice to use generic. It is better for caller to know about return type of method from its signature. Otherwise, we always have to look inside the implementation and see what will be the return type for each parameter combinations.

If we really want to use generic at least it would be better to look like
```
<T> T extract(T value, KNNVectorValuesIterator vectorValuesIterator)
```
But again, passing T value is strange because we only use it to determine return type...

Reply:
@heemin32 , VectorDataType is already getting passed in this function ...so at caller already knows what data types got passed, I don't foresee any confusion around implementation and generics use cases. 

I slightly disagree on this 

```
<T> T extract(T value, KNNVectorValuesIterator vectorValuesIterator)
```

Here are my reasons 

1. The use of <T> T in this method signature means that the type T is determined by the caller. However, without a clear relationship between T and the KNNVectorValuesIterator or the data being extracted, the method becomes ambiguous. There's no guarantee that the type of T matches the type of data the iterator can provide.
2. The parameter T value doesn't seem to serve a purpose in the context of an extraction operation. Typically, an extractor method would only require the source of data (in this case, KNNVectorValuesIterator) and any necessary metadata (like data type). The value parameter might confuse users, as it's unclear why they would provide an instance of T when they expect to receive one , on the other hand it is getting up fronted provided vector data type.

Reply:
`VectorDataType is already getting passed in this function` but who enforce it and which type does each VectorDataType map to? You know that because you are familiar with the inside code. 

Let's say I have a method `<T> T getSomthing(VectorDataType type, KNNVectorValuesIterator iterator, SomeotherClass class);` Can you guess what the T will be for each parameter?

Reply:
@heemin32  Yes this is how :- 

```
KNNVectorValues<float[]> floatVectorsWithMap = KNNVectorValuesFactory.getVectorValues(VectorType.FLOAT, docsWithFieldSet, floatVectorMap);
```